File: core/src/com/sheepsgohome/screens/GameplayDialog.java
package com.sheepsgohome.screens;

import com.badlogic.gdx.Game;
import com.badlogic.gdx.Gdx;
import com.badlogic.gdx.Screen;
import com.badlogic.gdx.audio.Sound;
import com.badlogic.gdx.graphics.GL20;
import com.badlogic.gdx.graphics.Texture;
import com.badlogic.gdx.scenes.scene2d.InputEvent;
import com.badlogic.gdx.scenes.scene2d.Stage;
import com.badlogic.gdx.scenes.scene2d.ui.Image;
import com.badlogic.gdx.scenes.scene2d.ui.Label;
import com.badlogic.gdx.scenes.scene2d.ui.Table;
import com.badlogic.gdx.scenes.scene2d.ui.TextButton;
import com.badlogic.gdx.scenes.scene2d.utils.ClickListener;
import com.badlogic.gdx.utils.viewport.StretchViewport;
import com.sheepsgohome.GameData;
import com.sheepsgohome.GameScreens;
import com.sheepsgohome.GameSounds;
import com.sheepsgohome.dialogs.NewBadgeDialog;

import java.util.Iterator;
import java.util.Vector;

import static com.badlogic.gdx.graphics.Texture.TextureFilter;
import static com.sheepsgohome.GameData.CAMERA_HEIGHT;
import static com.sheepsgohome.GameData.CAMERA_WIDTH;
import static com.sheepsgohome.GameData.Loc;
import static com.sheepsgohome.GameData.SOUND_ENABLED;
import static com.sheepsgohome.GameData.SOUND_VOLUME;
import static com.sheepsgohome.GameSkins.skin;

public class GameplayDialog implements Screen {


    private final int BADGES_COUNT = 12;

    private float BUTTON_WIDTH = 80;

    private GameDialogType type;

    private Stage stage;
    private Table table;

    private TextButton buttonRetry;
    private TextButton buttonNext;
    private TextButton buttonQuit;
    private Label title;

    private Texture sheep_texture;
    private Texture wolf_texture;
    private Texture texture;

    private Image imgSheep;
    private Image imgWolf;
    private Image bgImage;

    private Sound sound;

    private Vector<Texture> badges;

    public GameplayDialog(GameDialogType type) {
        float multiplier = 2;
        stage = new Stage(new StretchViewport(CAMERA_WIDTH * multiplier, CAMERA_HEIGHT * multiplier));
        table = new Table();

        sheep_texture = new Texture("sheep_success.png");
        sheep_texture.setFilter(TextureFilter.Linear, TextureFilter.Linear);


        switch (type) {
            case typeSheepFailed_Alpha:
                wolf_texture = new Texture("wolf_alpha_fail.png");
                break;
            case typeSheepFailed_Hungry:
                wolf_texture = new Texture("wolf_hungry_fail.png");
                break;
            default:
                wolf_texture = new Texture("wolf_fail.png");
                break;
        }
        wolf_texture.setFilter(TextureFilter.Linear, TextureFilter.Linear);

        badges = new Vector<Texture>();
        //fill vector
        for (int i = 0; i < BADGES_COUNT; i++) {
            Texture t = new Texture("badges/badge" + (i + 1) + ".png");
            t.setFilter(TextureFilter.Linear, TextureFilter.Linear);
            badges.add(t);
        }

        if (type != GameDialogType.typeSheepSucceeded) {
            buttonRetry = new TextButton(Loc.get("retry"), skin);
            imgWolf = new Image(wolf_texture);

            buttonRetry.addListener(new ClickListener() {
                @Override
                public void clicked(InputEvent event, float x, float y) {
                    ((Game) Gdx.app.getApplicationListener()).setScreen(GameScreens.gameplayClassicModeScreen);
                }
            });

        } else {
            buttonNext = new TextButton(Loc.get("next.level"), skin);
            imgSheep = new Image(sheep_texture);

            buttonNext.addListener(new ClickListener() {
                @Override
                public void clicked(InputEvent event, float x, float y) {
                    ((Game) Gdx.app.getApplicationListener()).setScreen(GameScreens.gameplayClassicModeScreen);
                }
            });

        }

        buttonQuit = new TextButton(Loc.get("quit"), skin);

        buttonQuit.addListener(new ClickListener() {
            @Override
            public void clicked(InputEvent event, float x, float y) {
                ((Game) Gdx.app.getApplicationListener()).setScreen(GameScreens.mainMenuScreen);
            }
        });

        //---------------------------------

        texture = new Texture("menu_background.png");
        texture.setFilter(TextureFilter.Linear, TextureFilter.Linear);

        bgImage = new Image(texture);
        bgImage.setWidth(CAMERA_WIDTH * multiplier);
        bgImage.setHeight(CAMERA_HEIGHT * multiplier);


        this.type = type;
        if (type != GameDialogType.typeSheepSucceeded) {
            title = new Label(Loc.get("sheep.has.been.caught"), skin, "menuTitle");
        } else {
            title = new Label(Loc.get("home.sweet.home"), skin, "menuTitle");
        }

        float fontScale = (CAMERA_WIDTH * multiplier - 20) / title.getPrefWidth();
        title.setFontScale(fontScale);
    }

    @Override
    public void show() {
        //table
//        table.setDebug(true);
        table.add(title).top().height(30).colspan(2).row();

        boolean showNewBadgeDialog = false;
        int badgeNo = 0;

        sound = null;

        if (type != GameDialogType.typeSheepSucceeded) {

            sound = GameSounds.soundWolfFailure;

            float mult = (CAMERA_WIDTH * 2) / imgWolf.getPrefWidth() * 0.95f;
            table.add(imgWolf).size(imgWolf.getWidth() * mult, imgWolf.getHeight() * mult).colspan(2).expand().row();

            table.add(buttonRetry).size(BUTTON_WIDTH, BUTTON_WIDTH / 2);
        } else {
            //has earned new badge?
            if ((badgeNo = getEarnedBadge(GameData.LEVEL)) > 0) {
                showNewBadgeDialog = true;
            }

            if (showNewBadgeDialog) {
                sound = GameSounds.soundNewBadge;
            } else {
                sound = GameSounds.soundSheepSuccess;
            }

            //display badges
            table.add(createBadgesTable(GameData.LEVEL)).expandX().colspan(2).row();

            GameData.LEVEL++;
            GameData.gamePreferences.putInteger("LEVEL", GameData.LEVEL);
            GameData.gamePreferences.flush();

            float mult = 0.30f;
            table.add(imgSheep).size(imgSheep.getWidth() * mult, imgSheep.getHeight() * mult).colspan(2).expand().row();

            table.add(buttonNext).size(BUTTON_WIDTH, BUTTON_WIDTH / 2);
        }


        table.add(buttonQuit).size(BUTTON_WIDTH, BUTTON_WIDTH / 2).row();
        table.setFillParent(true);

        stage.addActor(bgImage);
        stage.addActor(table);

        if (showNewBadgeDialog && badgeNo > 0) {
            new NewBadgeDialog(getBadgeName(badgeNo), badges.get(badgeNo - 1), skin, "dialog").show(stage);
        }

        Gdx.input.setInputProcessor(stage);

        if (sound != null && SOUND_ENABLED) {
            sound.play(SOUND_VOLUME);
        }
    }

    private String getBadgeName(int badgeNo) {
        switch (badgeNo) {
            case 1:
                return Loc.get("pasture");
            case 2:
                return Loc.get("threat.awareness");
            case 3:
                return Loc.get("alpha.defeater");
            case 4:
                return Loc.get("escapist");
            case 5:
                return Loc.get("agility");
            case 6:
                return Loc.get("tactician");
            case 7:
                return Loc.get("fearless");
            case 8:
                return Loc.get("sheeps.defiance");
            case 9:
                return Loc.get("limitless.courage");
            case 10:
                return Loc.get("ruthless.conspiracy");
            case 11:
                return Loc.get("wolf.apocalypse");
            case 12:
                return Loc.get("sheep.master");
            default:
                return "";
        }
    }

    private Table createBadgesTable(int level) {
        Table tab = new Table();

        int badgesCount = getBadgesCount(level);
        for (int i = 0; i < badgesCount; i++) {
            tab.add(new Image(badges.get(i))).size(28, 28).padRight(1);
            if (i == 5) {
                tab.row();
            }
        }

        return tab;
    }

    private int getBadgesCount(int level) {
        int ret = level / 10;
        return ret > BADGES_COUNT ? BADGES_COUNT : ret;
    }

    private int getEarnedBadge(int level) {
        if (level > 10 * BADGES_COUNT) {
            return 0;
        }

        return (level % 10) == 0 ? getBadgesCount(level) : 0;
    }

    @Override
    public void render(float delta) {
        Gdx.gl.glClearColor(0.5f, 0.5f, 0.5f, 1);
        Gdx.gl.glClear(GL20.GL_COLOR_BUFFER_BIT);
        stage.act();
        stage.draw();
    }

    @Override
    public void resize(int width, int height) {
        stage.getViewport().update(width, height, true);
    }

    @Override
    public void pause() {

    }

    @Override
    public void resume() {

    }

    @Override
    public void hide() {
        dispose();
    }

    @Override
    public void dispose() {
        stage.dispose();
        texture.dispose();
        sheep_texture.dispose();
        wolf_texture.dispose();

        //dispose vector textures
        Iterator<Texture> it = badges.iterator();
        while (it.hasNext()) {
            it.next().dispose();
        }
        badges.clear();
    }
}

File: core/src/com/sheepsgohome/screens/GameplayDialog.kt
package com.sheepsgohome.screens

import com.badlogic.gdx.Game
import com.badlogic.gdx.Gdx
import com.badlogic.gdx.Screen
import com.badlogic.gdx.audio.Sound
import com.badlogic.gdx.graphics.GL20
import com.badlogic.gdx.graphics.Texture
import com.badlogic.gdx.graphics.Texture.TextureFilter
import com.badlogic.gdx.scenes.scene2d.InputEvent
import com.badlogic.gdx.scenes.scene2d.Stage
import com.badlogic.gdx.scenes.scene2d.ui.Image
import com.badlogic.gdx.scenes.scene2d.ui.Label
import com.badlogic.gdx.scenes.scene2d.ui.Table
import com.badlogic.gdx.scenes.scene2d.ui.TextButton
import com.badlogic.gdx.scenes.scene2d.utils.ClickListener
import com.badlogic.gdx.utils.viewport.StretchViewport
import com.sheepsgohome.dialogs.NewBadgeDialog
import com.sheepsgohome.shared.GameData
import com.sheepsgohome.shared.GameData.CAMERA_HEIGHT
import com.sheepsgohome.shared.GameData.CAMERA_WIDTH
import com.sheepsgohome.shared.GameData.SOUND_ENABLED
import com.sheepsgohome.shared.GameData.SOUND_VOLUME
import com.sheepsgohome.shared.GameData.loc
import com.sheepsgohome.shared.GameScreens
import com.sheepsgohome.shared.GameSkins.skin
import com.sheepsgohome.shared.GameSounds
import java.util.*

class GameplayDialog(private val type: GameDialogType) : Screen {

    private val BADGES_COUNT = 12
    private val BUTTON_WIDTH = 80f

    private val stage: Stage
    private val table: Table

    private var buttonRetry: TextButton? = null
    private var buttonNext: TextButton? = null
    private val buttonQuit: TextButton
    private val title: Label

    private val sheep_texture: Texture
    private val wolf_texture: Texture
    private val texture: Texture

    private var imgSheep: Image? = null
    private var imgWolf: Image? = null
    private val bgImage: Image

    private var sound: Sound? = null

    private val badges: Vector<Texture>

    init {
        val multiplier = 2f
        stage = Stage(StretchViewport(CAMERA_WIDTH * multiplier, CAMERA_HEIGHT * multiplier))
        table = Table()

        sheep_texture = Texture("sheep_success.png")
        sheep_texture.setFilter(TextureFilter.Linear, TextureFilter.Linear)


        wolf_texture = when (type) {
            GameDialogType.typeSheepFailed_Alpha -> Texture("wolf_alpha_fail.png")
            GameDialogType.typeSheepFailed_Hungry -> Texture("wolf_hungry_fail.png")
            else -> Texture("wolf_fail.png")
        }
        wolf_texture.setFilter(TextureFilter.Linear, TextureFilter.Linear)

        badges = Vector()
        //fill vector
        for (i in 0..BADGES_COUNT - 1) {
            val t = Texture("badges/badge" + (i + 1) + ".png")
            t.setFilter(TextureFilter.Linear, TextureFilter.Linear)
            badges.add(t)
        }

        if (type !== GameDialogType.typeSheepSucceeded) {
            buttonRetry = TextButton(loc.get("retry"), skin)
            imgWolf = Image(wolf_texture)

            buttonRetry?.addListener(object : ClickListener() {
                override fun clicked(event: InputEvent?, x: Float, y: Float) {
                    (Gdx.app.applicationListener as Game).screen = GameScreens.gameplayClassicModeScreen
                }
            })

        } else {
            buttonNext = TextButton(loc.get("next.level"), skin)
            imgSheep = Image(sheep_texture)

            buttonNext?.addListener(object : ClickListener() {
                override fun clicked(event: InputEvent?, x: Float, y: Float) {
                    (Gdx.app.applicationListener as Game).screen = GameScreens.gameplayClassicModeScreen
                }
            })

        }

        buttonQuit = TextButton(loc.get("quit"), skin)

        buttonQuit.addListener(object : ClickListener() {
            override fun clicked(event: InputEvent?, x: Float, y: Float) {
                (Gdx.app.applicationListener as Game).screen = GameScreens.mainMenuScreen
            }
        })

        //---------------------------------

        texture = Texture("menu_background.png")
        texture.setFilter(TextureFilter.Linear, TextureFilter.Linear)

        bgImage = Image(texture)
        bgImage.width = CAMERA_WIDTH * multiplier
        bgImage.height = CAMERA_HEIGHT * multiplier

        title = when (type) {
            GameDialogType.typeSheepSucceeded -> Label(loc.get("home.sweet.home"), skin, "menuTitle")
            else -> Label(loc.get("sheep.has.been.caught"), skin, "menuTitle")
        }

        val fontScale = (CAMERA_WIDTH * multiplier - 20) / title.prefWidth
        title.setFontScale(fontScale)
    }

    override fun show() {
        //table
        table.add(title)
                .top()
                .height(30f)
                .colspan(2)
                .row()

        var showNewBadgeDialog = false
        var badgeNo = 0

        sound = null

        if (type !== GameDialogType.typeSheepSucceeded) {

            sound = GameSounds.soundWolfFailure

            imgWolf?.let {
                val mult = CAMERA_WIDTH * 2 / it.prefWidth * 0.95f

                table.add(it)
                    .size(it.width * mult, it.height * mult)
                    .colspan(2)
                    .expand()
                    .row()
            }

            table.add(buttonRetry).size(BUTTON_WIDTH, BUTTON_WIDTH / 2)

        } else {
            //has earned new badge?
            badgeNo = getEarnedBadge(GameData.LEVEL)
            if (badgeNo > 0) {
                showNewBadgeDialog = true
            }

            if (showNewBadgeDialog) {
                sound = GameSounds.soundNewBadge
            } else {
                sound = GameSounds.soundSheepSuccess
            }

            //display badges
            table.add(createBadgesTable(GameData.LEVEL)).expandX().colspan(2).row()

            GameData.levelUp()

            val mult = 0.30f

            imgSheep?.let {
                table.add(it)
                    .size(it.width * mult, it.height * mult)
                    .colspan(2)
                    .expand()
                    .row()
            }

            table.add(buttonNext).size(BUTTON_WIDTH, BUTTON_WIDTH / 2)
        }


        table.add(buttonQuit).size(BUTTON_WIDTH, BUTTON_WIDTH / 2).row()
        table.setFillParent(true)

        stage.addActor(bgImage)
        stage.addActor(table)

        if (showNewBadgeDialog && badgeNo > 0) {
            NewBadgeDialog(getBadgeName(badgeNo), badges[badgeNo - 1], skin, "dialog").show(stage)
        }

        Gdx.input.inputProcessor = stage

        if (SOUND_ENABLED) {
            sound?.play(SOUND_VOLUME)
        }
    }

    private fun getBadgeName(badgeNo: Int): String {
        when (badgeNo) {
            1 -> return loc.get("pasture")
            2 -> return loc.get("threat.awareness")
            3 -> return loc.get("alpha.defeater")
            4 -> return loc.get("escapist")
            5 -> return loc.get("agility")
            6 -> return loc.get("tactician")
            7 -> return loc.get("fearless")
            8 -> return loc.get("sheeps.defiance")
            9 -> return loc.get("limitless.courage")
            10 -> return loc.get("ruthless.conspiracy")
            11 -> return loc.get("wolf.apocalypse")
            12 -> return loc.get("sheep.master")
            else -> return ""
        }
    }

    private fun createBadgesTable(level: Int): Table {
        val tab = Table()

        val badgesCount = getBadgesCount(level)
        for (i in 0..badgesCount - 1) {
            tab.add(Image(badges[i])).size(28f, 28f).padRight(1f)
            if (i == 5) {
                tab.row()
            }
        }

        return tab
    }

    private fun getBadgesCount(level: Int): Int {
        val ret = level / 10
        return if (ret > BADGES_COUNT) BADGES_COUNT else ret
    }

    private fun getEarnedBadge(level: Int): Int {
        if (level > 10 * BADGES_COUNT) {
            return 0
        }

        return if (level % 10 == 0) getBadgesCount(level) else 0
    }

    override fun render(delta: Float) {
        Gdx.gl.glClearColor(0.5f, 0.5f, 0.5f, 1f)
        Gdx.gl.glClear(GL20.GL_COLOR_BUFFER_BIT)
        stage.act()
        stage.draw()
    }

    override fun resize(width: Int, height: Int) {
        stage.viewport.update(width, height, true)
    }

    override fun pause() {

    }

    override fun resume() {

    }

    override fun hide() {
        dispose()
    }

    override fun dispose() {
        stage.dispose()
        texture.dispose()
        sheep_texture.dispose()
        wolf_texture.dispose()

        //dispose vector textures
        val it = badges.iterator()
        while (it.hasNext()) {
            it.next().dispose()
        }
        badges.clear()
    }
}


